src/lib.rs

```diff
@@ -0,0 +1,631 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+use async_trait::async_trait;
+use prost::Message;
+use std::collections::HashMap;
+use std::sync::{
+    atomic::{AtomicU64, Ordering},
+    Arc, Mutex,
+};
+use std::time::Duration;
+use uprotocol_sdk::{
+    rpc::{RpcClient, RpcClientResult, RpcMapperError, RpcServer},
+    transport::{datamodel::UTransport, validator::Validators},
+    uprotocol::{
+        Data, UAttributes, UCode, UEntity, UMessage, UMessageType, UPayload, UPayloadFormat,
+        UPriority, UStatus, UUri, Uuid,
+    },
+    uri::{
+        serializer::{MicroUriSerializer, UriSerializer},
+        validator::UriValidator,
+    },
+};
+use zenoh::{
+    config::Config,
+    prelude::{r#async::*, Sample},
+    queryable::{Query, Queryable},
+    sample::AttachmentBuilder,
+    subscriber::Subscriber,
+};
+
+pub struct ZenohListener {}
+pub struct ULinkZenoh {
+    session: Arc<Session>,
+    subscriber_map: Arc<Mutex<HashMap<String, Subscriber<'static, ()>>>>,
+    queryable_map: Arc<Mutex<HashMap<String, Queryable<'static, ()>>>>,
+    query_map: Arc<Mutex<HashMap<String, Query>>>,
+    callback_counter: AtomicU64,
+}
+
+impl ULinkZenoh {
+    /// # Errors
+    /// Will return `Err` if unable to create Zenoh session
+    pub async fn new(config: Config) -> Result<ULinkZenoh, UStatus> {
+        let Ok(session) = zenoh::open(config).res().await else {
+            return Err(UStatus::fail_with_code(
+                UCode::Internal,
+                "Unable to open Zenoh session",
+            ));
+        };
+        Ok(ULinkZenoh {
+            session: Arc::new(session),
+            subscriber_map: Arc::new(Mutex::new(HashMap::new())),
+            queryable_map: Arc::new(Mutex::new(HashMap::new())),
+            query_map: Arc::new(Mutex::new(HashMap::new())),
+            callback_counter: AtomicU64::new(0),
+        })
+    }
+
+    fn to_zenoh_key_string(uri: &UUri) -> Result<String, UStatus> {
```

Comment:
In support of the uStreamer, we'll need some kind of special mode. Let me sketch out the idea here, but let's discuss more in tomorrow's meeting.

In uStreamer we need to be able to be able to subscribe to "any CEs which have a UAuthority filled in, and some valid Remote" so probably we need to modify this bit. I have a bit of a hack over [in a feature branch](https://github.com/PLeVasseur/uprotocol-rust-ulink-zenoh_zettascale/blob/feature/stand-in_for_uprotocol-rust-transport-mqtt/src/lib.rs#L116-L142) I'm using to keep moving on development of uStreamer.

So that'd mean that in:

`UTransport::register_listener()` we need a way to listen to all CEs where:
UAttributes.sink.authority has a remote &
UAttributes.type = UmessageTypeRequest

`RpcServer::register_rpc_listener()` we need a way to listen to all CEs where:
UAttributes.sink.authority has a remote &
UAttributes.type = UmessageTypePublish or UmessageTypeResponse

Reply:
Old comment. Refer to [this one instead](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/2/files#r1469865713) that fleshes the idea out more.